tensorflow/python/keras/engine/training.py

```diff
@@ -2082,6 +2083,12 @@ def save(self,
     save.save_model(self, filepath, overwrite, include_optimizer, save_format,
                     signatures, options, save_traces)
 
+  def __reduce__(self):
+    return pack_model(self)
+
+  def __deepcopy__(self):
+    return super().__deepcopy__()
```

Comment:
By default, `deepcopy` will check for `__deepcopy__` and fall back to `__reduce__` if the latter is implemented but not the former. This just forces the current behavior by implementing `__deepcopy__` but delegating it up the MRO.

Reply:
@jakirkham are there any cons to this approach? I think all we need for Dask and Joblib (by extension Scikit-Learn and others) is `__reduce__`, but it would be great to confirm.

Reply:
I'm now realizing that at least Sklearn _does_ use `deepcopy` directly ([here](https://github.com/scikit-learn/scikit-learn/blob/95119c13af77c76e150b753485c662b7c52a41a2/sklearn/base.py#L60))

@mihaimaruseac I'm going to take a couple of days to look into implementing this without delegating `__deepcopy__`. Sorry for pulling you in early.

Reply:
Using `__reduce__` instead of `__reduce_ex__` should be fine if we are just returning NumPy arrays. NumPy supports pickle protocol 5 internally (even with `pickle5` backport support for old Python version). So would be a good way to implement things if needed. Otherwise we need `__reduce_ex__` so we can wrap things with `PickleBuffer` ourselves. Hope that answers the question

Reply:
Yeah this was more asking about the `__deepcopy__` part, in particular I was wondering if Dask or anything in your world uses `__deepcopy__`. But I think I resolved this in https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-799786980